### PR TITLE
Support random state and CV

### PR DESCRIPTION
**Pull request by BenjaminBossan**
Add support for `RandomState` and `Generator`. Unfortunately, the numpy docs were a bit sparse when it comes to how to save and load those, I think I got it right but please check. Especially, do you think the new test is rigorous enough?

Regarding cross validators, ideally we would have a something like `estimator_checks` but for cross validators instead. I didn't find anything (LMK if there is) so for now I just tested a couple of cross validators but not all.

As you can see, the cross validators just use `BaseEstimator_get_state` and `BaseEstimator_get_instance` under the hood. I assume that many of the failing tests can also be fixed by falling back on those two functions when we encounter an sklearn class which is not taken care of (if it's not Cython). This makes me wonder if we could/should have something like a module level fallback. Since we control the dispatcher, it should be possible to implement this.

As for the tests, I did not try to save and load random states and cross validators as "first class" objects but always as attributes on a `BaseEstimator`. I was not sure if we want to support them as "first class" objects or not.

Finally, I fixed the bug where sparse matrices were stored on the project root.

Please review @adrinjalali. Locally, this gives me `50 failed, 360 passed`.

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
Side note: I skimmed the remaining failing tests and they can be categorized into the following failure cases:

### Calling estimator incorrectly

E.g. `TypeError: PLSRegression.fit() got an unexpected keyword argument 'y'` or calling `CountVectorizer` with floats instead of strings.

### Custom sklearn objects

E.g. `TypeError: Getting the state of type <class 'sklearn.linear_model._sgd_fast.Hinge'> is not supported yet`. This may be solved if we add fallbacks, otherwise we would probably need a special dispatcher for each of those.

### Failing equality checks for custom objects

Since by default, two objects instances are not identical, even if all their attributes are.

### Scipy ufuncs

Since they look like numpy ufuncs

### Incorrect casting

Custom dict types are treated as dict (`assert <class 'sklearn.covariance._graph_lasso._DictWithDeprecatedKeys'> == <class 'dict'>`), json dicts not supporting non-string keys (`AssertionError: assert {1} == {'1'}`).

File: skops/io/_numpy.py [hunk 1]
```diff
@@ -47,6 +48,48 @@ def ndarray_get_instance(state, src):
...
 
+def random_state_get_state(obj, dst):
+    content = get_state(obj.get_state(legacy=False), dst)
+    res = {
+        "__class__": obj.__class__.__name__,
+        "__module__": get_module(type(obj)),
+        "type": "numpy",
+        "content": content,
+    }
+    return res
```
Comment (adrinjalali) [inline skops/io/_numpy.py:59]: @rgommers would this be enough to save the state of a random state? Same with a generator bellow.

This is in the context of saving sklearn-compatible models while not using pickle.

File: skops/io/_scipy.py [hunk 2]
```diff
@@ -11,7 +12,7 @@ def sparse_matrix_get_state(obj, dst):
         "__module__": get_module(type(obj)),
     }
 
-    f_name = f"{uuid4()}.npz"
+    f_name = Path(obj) / f"{uuid4()}.npz"
```
Comment (adrinjalali) [inline skops/io/_scipy.py]: could you push this one directly to the repo please?

Also, wondering why this wasn't caught by tests.

Comment (BenjaminBossan) [reply skops/io/_scipy.py]: Done.

Errors are hard to spot with the amount of still failing test, maybe we just missed it.

File: skops/io/_numpy.py (hunk 1 above)
Comment (rgommers) [reply skops/io/_numpy.py:59]: I _think_ that `bit_generator.state`, which seems to be used here, should be enough (hard to be certain without trying of course). And I'm assuming that, like pickle, the same numpy version is used to load the data in again.

Comment (adrinjalali) [reply skops/io/_numpy.py:59]: Oh that's interesting. For sklearn we need to use the same sklearn version to load things since things might move around. Should we be imposing the same restriction/warning for numpy/scipy? I thought `npy` format was more stable and we could safely kinda ignore the numpy version @rgommers .

sklearn gives a warning if the user tries to load using a different version, but I haven't seen numpy do that.

Comment (BenjaminBossan) [reply skops/io/_numpy.py:59]: > hard to be certain without trying of course

Would you say the [test](https://github.com/BenjaminBossan/skops/blob/aae23aa693f496d30c6615b411d86c505f9ff297/skops/io/tests/test_persist.py#L292-L303) is sufficient or could it be more comprehensive to cover some edge cases (say, multiprocessing or something like that)?

Theoretically, we could also test different numpy versions on CI but I think that would blow up our test matrix size, so it's unlikely.

Testing _different_ numpy versions for saving and loading seems quite hard to achieve. It would be good to know if numpy gives any guarantees there, but I didn't see anything in the docs.

Comment (rgommers) [reply skops/io/_numpy.py:59]: > I thought `npy` format was more stable and we could safely kinda ignore the numpy version @rgommers .

The `npy` format is indeed stable and versioned, and does not depend on the `numpy` version. However, (a) I didn't/don't see that used here, and (b) you seem to rely on the `BitGenerator` state, which AFAIK doesn't have guarantees as strong as the `.npy` format. xref https://numpy.org/neps/nep-0019-rng-policy.html for that

**Comment by BenjaminBossan** [comment]
After latest merge, I get `24 failed, 386 passed` vs `32 failed, 368 passed` on original branch when testing locally.
